java.lang.NullPointerException occurred when PostgreSQL describing insert returning composite type

## Bug Report

### Which version of ShardingSphere did you use?

master - 440536943ccad254425c25aadb1b491ffd9d1670

This may affect all versions before.

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy PostgreSQL

### Expected behavior

No error occurred.

### Actual behavior

Client:
```
Exception in thread "main" org.postgresql.util.PSQLException: ERROR: java.lang.NullPointerException
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:315)
	at org.postgresql.jdbc.PgPreparedStatement.getMetaData(PgPreparedStatement.java:1108)
```

Proxy:
```
[ERROR] 2022-12-26 16:27:41.316 [Connection-1-ThreadExecutor] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
java.lang.NullPointerException: null
	at java.base/java.util.Objects.requireNonNull(Objects.java:208)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.describe.PostgreSQLComDescribeExecutor.describeReturning(PostgreSQLComDescribeExecutor.java:201)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.describe.PostgreSQLComDescribeExecutor.lambda$describeInsertStatementByDatabaseMetaData$0(PostgreSQLComDescribeExecutor.java:139)
	at java.base/java.util.Optional.map(Optional.java:260)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.describe.PostgreSQLComDescribeExecutor.describeInsertStatementByDatabaseMetaData(PostgreSQLComDescribeExecutor.java:139)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.describe.PostgreSQLComDescribeExecutor.tryDescribePreparedStatement(PostgreSQLComDescribeExecutor.java:121)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.describe.PostgreSQLComDescribeExecutor.describePreparedStatement(PostgreSQLComDescribeExecutor.java:113)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.describe.PostgreSQLComDescribeExecutor.execute(PostgreSQLComDescribeExecutor.java:97)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.PostgreSQLAggregatedCommandExecutor.execute(PostgreSQLAggregatedCommandExecutor.java:41)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:110)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:77)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

No rule configured in logic database.

```sql
create table t (id int, val varchar);
```

### Example codes for reproduce this issue (such as a github link).

```java
PreparedStatement preparedStatement = connection.prepareStatement("insert into t (val) values ('') returning t");
ResultSetMetaData metaData = preparedStatement.getMetaData();
metaData.getColumnLabel(1);
```

--- a/db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/PostgreSQLColumnType.java
+++ b/db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/PostgreSQLColumnType.java
@@ -176,6 +176,8 @@ public enum PostgreSQLColumnType implements BinaryColumnType {
         JDBC_TYPE_AND_COLUMN_TYPE_MAP.put(Types.OTHER, POSTGRESQL_TYPE_JSON);
         JDBC_TYPE_AND_COLUMN_TYPE_MAP.put(Types.SQLXML, POSTGRESQL_TYPE_XML);
         JDBC_TYPE_AND_COLUMN_TYPE_MAP.put(Types.BOOLEAN, POSTGRESQL_TYPE_BOOL);
+        // TODO Temporary solution for https://github.com/apache/shardingsphere/issues/22522
+        JDBC_TYPE_AND_COLUMN_TYPE_MAP.put(Types.STRUCT, POSTGRESQL_TYPE_VARCHAR);
     }
     
     /**

--- a/proxy/frontend/postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/describe/PostgreSQLComDescribeExecutor.java
+++ b/proxy/frontend/postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/describe/PostgreSQLComDescribeExecutor.java
@@ -196,9 +196,10 @@ public final class PostgreSQLComDescribeExecutor implements CommandExecutor {
                         .forEach(result::add);
             }
             if (each instanceof ColumnProjectionSegment) {
-                String columnName = ((ColumnProjectionSegment) each).getColumn().getIdentifier().getValue();
-                ShardingSphereColumn column = columnsOfTable.getOrDefault(columnName, caseInsensitiveColumnsOfTable.get(columnName));
-                String alias = ((ColumnProjectionSegment) each).getAlias().orElseGet(column::getName);
+                ColumnProjectionSegment segment = (ColumnProjectionSegment) each;
+                String columnName = segment.getColumn().getIdentifier().getValue();
+                ShardingSphereColumn column = columnsOfTable.getOrDefault(columnName, caseInsensitiveColumnsOfTable.getOrDefault(columnName, generateDefaultColumn(segment)));
+                String alias = segment.getAlias().orElseGet(column::getName);
                 result.add(new PostgreSQLColumnDescription(alias, 0, column.getDataType(), estimateColumnLength(column.getDataType()), ""));
             }
             if (each instanceof ExpressionProjectionSegment) {
@@ -206,6 +207,10 @@ public final class PostgreSQLComDescribeExecutor implements CommandExecutor {
             }
         }
         return new PostgreSQLRowDescriptionPacket(result);
+    }
+    
+    private ShardingSphereColumn generateDefaultColumn(final ColumnProjectionSegment segment) {
+        return new ShardingSphereColumn(segment.getColumn().getIdentifier().getValue(), Types.VARCHAR, false, false, false, true, false);
     }
     
     private PostgreSQLColumnDescription convertExpressionToDescription(final ExpressionProjectionSegment expressionProjectionSegment) {

--- a/proxy/frontend/postgresql/src/test/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/describe/PostgreSQLComDescribeExecutorTest.java
+++ b/proxy/frontend/postgresql/src/test/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/describe/PostgreSQLComDescribeExecutorTest.java
@@ -255,7 +255,7 @@ public final class PostgreSQLComDescribeExecutorTest extends ProxyContextRestore
         final String statementId = "S_2";
         when(packet.getName()).thenReturn(statementId);
         String sql = "insert into t_order (k, c, pad) values (?, ?, ?) "
-                + "returning id, id alias_id, 'anonymous', 'OK' literal_string, 1 literal_int, 4294967296 literal_bigint, 1.1 literal_numeric, t_order.*";
+                + "returning id, id alias_id, 'anonymous', 'OK' literal_string, 1 literal_int, 4294967296 literal_bigint, 1.1 literal_numeric, t_order.*, t_order";
         SQLStatement sqlStatement = SQL_PARSER_ENGINE.parse(sql, false);
         List<PostgreSQLColumnType> parameterTypes = new ArrayList<>(sqlStatement.getParameterCount());
         for (int i = 0; i < sqlStatement.getParameterCount(); i++) {
@@ -276,7 +276,7 @@ public final class PostgreSQLComDescribeExecutorTest extends ProxyContextRestore
         DatabasePacket<?> actualRowDescriptionPacket = actualPacketsIterator.next();
         assertThat(actualRowDescriptionPacket, is(instanceOf(PostgreSQLRowDescriptionPacket.class)));
         List<PostgreSQLColumnDescription> actualColumnDescriptions = new ArrayList<>(getColumnDescriptionsFromPacket((PostgreSQLRowDescriptionPacket) actualRowDescriptionPacket));
-        assertThat(actualColumnDescriptions.size(), is(11));
+        assertThat(actualColumnDescriptions.size(), is(12));
         assertThat(actualColumnDescriptions.get(0).getColumnName(), is("id"));
         assertThat(actualColumnDescriptions.get(0).getTypeOID(), is(PostgreSQLColumnType.POSTGRESQL_TYPE_INT4.getValue()));
         assertThat(actualColumnDescriptions.get(0).getColumnLength(), is(4));
@@ -310,6 +310,9 @@ public final class PostgreSQLComDescribeExecutorTest extends ProxyContextRestore
         assertThat(actualColumnDescriptions.get(10).getColumnName(), is("pad"));
         assertThat(actualColumnDescriptions.get(10).getTypeOID(), is(PostgreSQLColumnType.POSTGRESQL_TYPE_CHAR.getValue()));
         assertThat(actualColumnDescriptions.get(10).getColumnLength(), is(-1));
+        assertThat(actualColumnDescriptions.get(11).getColumnName(), is("t_order"));
+        assertThat(actualColumnDescriptions.get(11).getTypeOID(), is(PostgreSQLColumnType.POSTGRESQL_TYPE_VARCHAR.getValue()));
+        assertThat(actualColumnDescriptions.get(11).getColumnLength(), is(-1));
     }
     
     @SuppressWarnings("unchecked")

--- a/proxy/frontend/postgresql/src/test/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/describe/PostgreSQLComDescribeExecutorTest.java
+++ b/proxy/frontend/postgresql/src/test/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/describe/PostgreSQLComDescribeExecutorTest.java
@@ -255,7 +255,7 @@ public final class PostgreSQLComDescribeExecutorTest extends ProxyContextRestore
         final String statementId = "S_2";
         when(packet.getName()).thenReturn(statementId);
         String sql = "insert into t_order (k, c, pad) values (?, ?, ?) "
-                + "returning id, id alias_id, 'anonymous', 'OK' literal_string, 1 literal_int, 4294967296 literal_bigint, 1.1 literal_numeric, t_order.*, t_order";
+                + "returning id, id alias_id, 'anonymous', 'OK' literal_string, 1 literal_int, 4294967296 literal_bigint, 1.1 literal_numeric, t_order.*, t_order, t_order alias_t_order";
         SQLStatement sqlStatement = SQL_PARSER_ENGINE.parse(sql, false);
         List<PostgreSQLColumnType> parameterTypes = new ArrayList<>(sqlStatement.getParameterCount());
         for (int i = 0; i < sqlStatement.getParameterCount(); i++) {
@@ -275,8 +275,12 @@ public final class PostgreSQLComDescribeExecutorTest extends ProxyContextRestore
         verify(mockPayload, times(2)).writeInt4(18);
         DatabasePacket<?> actualRowDescriptionPacket = actualPacketsIterator.next();
         assertThat(actualRowDescriptionPacket, is(instanceOf(PostgreSQLRowDescriptionPacket.class)));
-        List<PostgreSQLColumnDescription> actualColumnDescriptions = new ArrayList<>(getColumnDescriptionsFromPacket((PostgreSQLRowDescriptionPacket) actualRowDescriptionPacket));
-        assertThat(actualColumnDescriptions.size(), is(12));
+        assertRowDescriptions((PostgreSQLRowDescriptionPacket) actualRowDescriptionPacket);
+    }
+    
+    private void assertRowDescriptions(final PostgreSQLRowDescriptionPacket actualRowDescriptionPacket) {
+        List<PostgreSQLColumnDescription> actualColumnDescriptions = new ArrayList<>(getColumnDescriptionsFromPacket(actualRowDescriptionPacket));
+        assertThat(actualColumnDescriptions.size(), is(13));
         assertThat(actualColumnDescriptions.get(0).getColumnName(), is("id"));
         assertThat(actualColumnDescriptions.get(0).getTypeOID(), is(PostgreSQLColumnType.POSTGRESQL_TYPE_INT4.getValue()));
         assertThat(actualColumnDescriptions.get(0).getColumnLength(), is(4));
@@ -313,6 +317,9 @@ public final class PostgreSQLComDescribeExecutorTest extends ProxyContextRestore
         assertThat(actualColumnDescriptions.get(11).getColumnName(), is("t_order"));
         assertThat(actualColumnDescriptions.get(11).getTypeOID(), is(PostgreSQLColumnType.POSTGRESQL_TYPE_VARCHAR.getValue()));
         assertThat(actualColumnDescriptions.get(11).getColumnLength(), is(-1));
+        assertThat(actualColumnDescriptions.get(12).getColumnName(), is("alias_t_order"));
+        assertThat(actualColumnDescriptions.get(12).getTypeOID(), is(PostgreSQLColumnType.POSTGRESQL_TYPE_VARCHAR.getValue()));
+        assertThat(actualColumnDescriptions.get(12).getColumnLength(), is(-1));
     }
     
     @SuppressWarnings("unchecked")

